_posts/2018-01-07-building-serverless-radiator.md

```diff
@@ -0,0 +1,225 @@
+            "AlarmName": "product-name-service-500",
+            "StateValue": "OK"
+        }
+    ],
+    "pipelines": [
+        {
+            "name": "ProjectInfraBuildPipeline",
+            "currentStatus": "Succeeded"
+        }
+    ],
+    "metrics": [
+        {
+            "name": "RDS CPU Utilization",
+            "statistics": "Average",
+            "unit": "Percent",
+            "result": 7.004
+        }
+    ]
+}
+```
+
+The main handler is therefore quite simple. There is no need to use any parameters because we return always all the pipelines and alarms.   
+
+```python
+def status(event, context):
+    result = {
+        "alarms":get_alarms(),
+        "pipelines" : get_pipelines(),
+        "metrics" : get_metrics()}
+    return {"body": json.dumps(result)}
+```
+
+Getting pipeline statuses and alarms is quite straightforward. The alarms can be fetched with one API call. Pipelines need to be first fetched and then current statuses can be checked one by one. For the radiator I also formatted the result a little bit.
+
+Metrics cannot be fetched with one call because AWS has tons of different kind metrics available. Currently I make a Python data structure which describes what kind of metric I want to expose. The next code example gets average RDS CPU utilization of last 10 minutes.
+
+```python
+metrics= [{'name': 'RDS CPU Utilization',
+            'request': {'Namespace':'AWS/RDS',
+                         'MetricName':'CPUUtilization',
+                         'StartTime': datetime.utcnow() - timedelta(minutes=10),
+                         'EndTime': datetime.now() ,
+                         'Period': 600,
+                         'Statistics':['Average'],
+                         'Unit':'Percent'},
+            'statistics': 'Average',
+            'unit': 'Percent'}]
+```            
+
+Lambda can be installed by hand via AWS Console but automatized installing is much more preferable. Fortunately [Serverless Framework](https://serverless.com/) supports also deploying Python code. 
+
+With Serverless installing the Radiator Exposer is easy. Just run ``serverless deploy``. Serverless  runs for a while and returns a new endpoint and an api key necessary for authentication. 
+
+The Radiator exposer is a free software which can be downloaded from [https://github.com/hjhamala/radiator-exposer](https://github.com/hjhamala/radiator-exposer).
+
+
+## Building a radiator
+
+Getting information from an account is not enough. There should be some fancy and visual way to show it. This can be made also via serverless approach. 
+
+For the radiator I chose to make Clojure Lambda which polls endpoints and generates plain old html response with automatic refresh. Because AWS Lambda supports Java is Clojure therefore supported. For generating necessary Java classes I used [Lambada](https://github.com/uswitch/lambada) library. The radiator should be run also locally so I wrote also a simple web server for local usage.
```

Comment:
"make *a* Clojure Lambda"

"generates *a* plain old *HTML* response"

"…supports Java, it also supports Clojure."

"For generating *the* necessary Java classes, I used *the*…"

"You should also be able to run the radiator locally, so I also wrote a simple web server for local use."